tiledb/api/src/query/read/mod.rs

```diff
@@ -374,6 +384,262 @@ pub trait ReadQueryBuilder<'data>: QueryBuilder {
     }
 }
 
+pub struct AggregateBuilder<B, T> {
+    base: B,
+    agg_str: CString,
+    attr_str: Option<CString>,
+    attr_type: PhantomData<T>,
+}
+
+pub struct AggregateReader<Q, T> {
+    base: Q,
+    agg_str: CString,
+    _attr_str: Option<CString>,
+    data : T,
+    data_size : u64
+}
+
+impl<B, T> QueryBuilder for AggregateBuilder<B, T>
+where
+    B: QueryBuilder,
+    T : Default,
+{
+    type Query = AggregateReader<B::Query, T>;
+
+    fn base(&self) -> &BuilderBase {
+        &self.base.base()
+    }
+
+    fn build(self) -> Self::Query {
+        AggregateReader::<B::Query, T> {
+            base: self.base.build(),
+            agg_str: self.agg_str,
+            _attr_str: self.attr_str,
+            data: T::default(),
+            data_size : mem::size_of::<T>() as u64
+        }
+    }
+}
+
+impl<Q, T> Query for AggregateReader<Q, T>
+where
+    Q: Query,
+{
+    fn base(&self) -> &QueryBase {
+        &self.base.base()
+    }
+
+    fn finalize(self) -> TileDBResult<Array>
+    where
+        Self: Sized,
+    {
+        self.base.finalize()
+    }
+}
+
+impl<Q, T> ReadQuery for AggregateReader<Q, T>
+where
+    Q: ReadQuery,
+    T : Copy
+{
+    type Intermediate = ();
+    type Final = T;
+
+    fn step(
+        &mut self,
+    ) -> TileDBResult<ReadStepOutput<Self::Intermediate, Self::Final>> {
+        // Register the data buffer (set data buffer)
+        let context = self.base().context();
+        let cquery = **self.base().cquery();
+        let location_ptr = &mut self.data as *mut T;
+        let c_bufptr = location_ptr as *mut std::ffi::c_void;
+        let c_sizeptr = &mut self.data_size as *mut u64;
+        let agg_str: &CString = &self.agg_str;
+        let agg_c_ptr = agg_str.as_c_str().as_ptr() as *const i8;
+
+        context.capi_call(|ctx| unsafe {
+            ffi::tiledb_query_set_data_buffer(
+                ctx, cquery, agg_c_ptr, c_bufptr, c_sizeptr,
+            )
+        })?;
+
+        let base_result = self.base.step()?;
+
+        // Run the query in a loop until you get the final result
+        let return_val = match base_result {
+            ReadStepOutput::Final(_) => self.data,
+            ReadStepOutput::Intermediate(_) => unreachable!(),
+            ReadStepOutput::NotEnoughSpace => unreachable!(),
+        };
+
+        Ok(ReadStepOutput::Final(return_val))
+    }
+}
+
+#[derive(PartialEq)]
+pub enum AggregateType {
+    Count,
+    Max,
+    Mean,
+    Min,
+    NullCount,
+    Sum,
+}
+
+pub trait AggregateBuilderTrait: QueryBuilder {
+    fn apply_aggregate<T>(
+        self,
+        agg_type: AggregateType,
+        name: Option<String>,
```

Comment:
It's a little awkward since Count is the only one that doesn't require a name.  I think it would be better to put the field name inside of the enum, e.g.
```
enum AggregateType {
    Count,
    Max(String),
    ...
}
```
instead of a type and separate option.

Reply:
done!